esmvalcore/esgf/_download.py

```diff
@@ -0,0 +1,450 @@
+
+        Raises
+        ------
+        DownloadError:
+            Raised if downloading the file failed.
+
+        Returns
+        -------
+        Path
+            The path where the file will be located after download.
+        """
+        local_file = self.local_file(dest_folder)
+        if local_file.exists():
+            logger.debug("Skipping download of existing file %s", local_file)
+            return local_file
+
+        os.makedirs(local_file.parent, exist_ok=True)
+
+        errors = {}
+        for url in sort_hosts(self.urls):
+            try:
+                self._download(local_file, url)
+            except (DownloadError,
+                    requests.exceptions.RequestException) as error:
+                logger.debug("Not able to download %s. Error message: %s", url,
+                             error)
+                errors[url] = error
+                log_error(url)
+            else:
+                break
+
+        if not local_file.exists():
+            raise DownloadError(
+                f"Failed to download file {local_file}, errors:"
+                "\n" + "\n".join(f"{url}: {errors[url]}" for url in errors))
+
+        return local_file
+
+    @staticmethod
+    def _tmp_local_file(local_file):
+        """Return the path to a temporary local file for downloading to."""
+        with NamedTemporaryFile(prefix=f"{local_file}.") as tmp_file:
+            return Path(tmp_file.name)
+
+    def _download(self, local_file, url):
+        """Download file from a single url."""
+        idx = self.urls.index(url)
+        checksum_type, checksum = self._checksums[idx]
+        if checksum_type is None:
+            hasher = None
+        else:
+            hasher = hashlib.new(checksum_type)
+
+        tmp_file = self._tmp_local_file(local_file)
+
+        logger.debug("Downloading %s to %s", url, tmp_file)
+        start_time = datetime.datetime.now()
+        response = requests.get(url,
+                                stream=True,
+                                timeout=TIMEOUT,
+                                cert=get_credentials())
+        response.raise_for_status()
+        with tmp_file.open("wb") as file:
+            for chunk in response.iter_content(chunk_size=None):
+                if hasher is not None:
+                    hasher.update(chunk)
+                file.write(chunk)
+
+        duration = datetime.datetime.now() - start_time
+
+        if hasher is None:
+            logger.warning(
+                "No checksum available, unable to check data"
+                " integrity for %s, ", url)
+        else:
+            local_checksum = hasher.hexdigest()
+            if local_checksum != checksum:
+                raise DownloadError(
+                    f"Wrong {checksum_type} checksum for file {tmp_file},"
+                    f" downloaded from {url}: expected {checksum}, but got"
+                    f" {local_checksum}. Try downloading the file again.")
```

Comment:
Maybe in a follow up pull request, for now this should be enough. If you want another try, you can just re-run the recpie.

Reply:
bah humbug, after having downloaded all those files and stumbled only on an unresponsive node for exactly 5s :grin: OK fine, I guess we should think about it since 1. it'll spend a bit of time downloading, then it'll spend a bit of time finding all those files at rerun, nit picky users will be complaining